pkg/network/netlink/socket.go

```diff
@@ -0,0 +1,260 @@
+)
+
+var _ netlink.Socket = &Socket{}
+var errNotImplemented = errors.New("not implemented")
+
+// Socket is an implementation of netlink.Socket (github.com/mdlayher/netlink)
+// It's mostly a copy of the original implementation (netlink.conn) with a few optimizations:
+// * We don't MSG_PEEK as we use a pre-allocated buffer large enough to fit any netlink message;
+// * We use a buffer pool for the message data;
+// * We remove all the synchronization & go-channels cruft and bring it upstream in a cheaper/simpler way (Consumer)
+type Socket struct {
+	fd   *os.File
+	pid  uint32
+	conn syscall.RawConn
+
+	// A 32KB buffer which we use for polling the socket.
+	// Since a netlink message can't exceed that size
+	// (in *theory* they can be as large as 4GB (u32), but see link below)
+	// we can avoid message peeks and and essentially cut recvmsg syscalls by half
+	// which is currently a perf bottleneck in certain workloads.
+	// https://www.spinics.net/lists/netdev/msg431592.html
+	recvbuf []byte
+}
+
+// NewSocket creates a new NETLINK socket
+func NewSocket() (*Socket, error) {
+	fd, err := unix.Socket(
+		unix.AF_NETLINK,
+		unix.SOCK_RAW|unix.SOCK_CLOEXEC,
+		unix.NETLINK_NETFILTER,
+	)
+
+	if err != nil {
+		return nil, err
+	}
+
+	err = unix.SetNonblock(fd, true)
+	if err != nil {
+		syscall.Close(fd)
+		return nil, err
+	}
+
+	err = unix.Bind(fd, &unix.SockaddrNetlink{Family: unix.AF_NETLINK})
+	if err != nil {
+		syscall.Close(fd)
+		return nil, os.NewSyscallError("bind", err)
+	}
+
+	addr, err := unix.Getsockname(fd)
+	if err != nil {
+		syscall.Close(fd)
+		return nil, os.NewSyscallError("getsockname", err)
+	}
+
+	pid := addr.(*unix.SockaddrNetlink).Pid
+	file := os.NewFile(uintptr(fd), "netlink")
+
+	conn, err := file.SyscallConn()
+	if err != nil {
+		file.Close()
+		return nil, err
+	}
+
+	socket := &Socket{
+		fd:      file,
+		pid:     pid,
+		conn:    conn,
+		recvbuf: make([]byte, 32*1024),
+	}
+	return socket, nil
+}
+
+// Send a netlink.Message
+func (s *Socket) Send(m netlink.Message) error {
+	b, err := m.MarshalBinary()
+	if err != nil {
+		return err
+	}
+
+	addr := &unix.SockaddrNetlink{
+		Family: unix.AF_NETLINK,
+	}
+
+	ctrlErr := s.conn.Write(func(fd uintptr) bool {
+		err = unix.Sendmsg(int(fd), b, nil, addr, 0)
+		return ready(err)
+	})
+	if ctrlErr != nil {
+		return ctrlErr
+	}
+
+	return err
+}
+
+// Receive is not implemented. See ReceiveInto
+func (s *Socket) Receive() ([]netlink.Message, error) {
+	return nil, errNotImplemented
+}
+
+// ReceiveInto reads one or more netlink.Messages off the socket
+func (s *Socket) ReceiveInto(b []byte) ([]netlink.Message, error) {
+	n, err := s.recvmsg(s.recvbuf, 0)
+	if err != nil {
+		return nil, os.NewSyscallError("recvmsg", err)
+	}
+
+	n = nlmsgAlign(n)
+	if n > len(b) {
+		b = make([]byte, n)
+	}
+	copy(b, s.recvbuf[:n])
+
+	raw, err := syscall.ParseNetlinkMessage(b[:n])
+	if err != nil {
+		return nil, err
+	}
+
+	msgs := make([]netlink.Message, 0, len(raw))
```

Comment:
I think that's a fair point, but I'm not a big fan of the idea since that this would mean that a single go-routine is responsible for all of the heavy-lifting of doing the socket polling and decoding the messages.

Besides that, I took a alloc profile and this is responsible for the 0.65% of the allocations, so honestly I don't think it's an optimization worth the hassle at this point.